Added Checks for the Length of Arguments List

BowTiedWoo opened the pull request:
Added checks in `traverse` functions throughout the code in order to verify that the number of arguments provided at compile-time is the expected one for the given function.

Reference issue: https://github.com/stacks-network/clarity-wasm/issues/159

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/stacks-network/clarity-wasm?pullRequest=533) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/stacks-network/clarity-wasm?pullRequest=533) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/stacks-network/clarity-wasm?pullRequest=533) it.</sub>

Acaccia reviewed (changes requested):
I would go for `Generator::TypeError` for all these. Giving one too many or too few arguments to a function doesn't respect the signature of the function, which I would qualify as a typing error.
What do you think?

`InternalError` should be reserved for the cases where something unexpected happens with the compilation process.

csgui commented:
> I would go for `Generator::TypeError` for all these...

@Acaccia I would expect a `Generator::TypeError` for an argument case if the function expects an `int` and I am passing a `uint`, for instance. 

What you think in adding a new ` GeneratorError::ArgumentLenghtError` for the cases in this PR? That should be simple.

csgui reviewed (changes requested):
Added a comment to discussion.

BowTiedWoo commented:
Thank you for the feedback @Acaccia, @csgui !
I also think they are a bit different and `GeneratorError::ArgumentLenghtError` is more specific. I will create a new error type for that, if it's alright, as it doesn't add to much complexity on my end.

csgui commented on clar2wasm/src/words/bindings.rs:
```diff
@@ -18,6 +18,13 @@ impl ComplexWord for Let {
         _expr: &SymbolicExpression,
         args: &[SymbolicExpression],
     ) -> Result<(), GeneratorError> {
+        if args.len() < 2 {
+            return Err(GeneratorError::ArgumentLengthError(format!(
+                "let expected at least 2 arguments, got {}",
+                args.len()
+            )));
+        };
```
Can you please cover these changes with tests?

BowTiedWoo replied:
I’m updating them now accordingly to the review.

BowTiedWoo replied:
Done, I've added the tests.

Acaccia commented:
@chris We can always add more types of errors, our current error handling is left to be desired.

For example, in #516, I had to use a disgusting regex to match the string representation of the error instead of dealing directly with an error type. That was subpar.

I'm not entirely sure what this type would add compared to a type error here (whatever way I look at this, for me passing the wrong number of arguments to a function is a type error). But here you are the one who dealt with adapting our errors to the interpreter's, so I will defer to your judgement here.

csgui commented on clar2wasm/src/words/blockinfo.rs:
```diff
@@ -227,6 +248,16 @@ mod tests {
 
     //- Block Info
 
+    #[test]
+    fn get_block_info_less_than_two_args() {
+        crosscheck_expect_failure("(get-block-info? id-header-hash)");
+    }
```
`crosscheck_expect_failure` is somewhat deprecated. We should avoid using it since we don’t know which error is actually being caught by that function. Please check comments [here](https://github.com/stacks-network/clarity-wasm/blob/main/clar2wasm/src/tools.rs#L688-L693).

Here’s a good example of that. The error being thrown is probably not the expected `ArgumentError`.

Consider using other variations of `crosscheck` functions throught this PR.

BowTiedWoo replied:
I updated the tests so they do not include `crosscheck_expect_failure`.

They are now using the `evaluate` function.

csgui commented:
Just adding more context: not sure if the `ArgumentLengthError` will be caught since a compile error is thrown before the traverse step.

@Acaccia what is your take on that? It seems that this PR should be handled after the work on compile-time error handling, issue #421.

Acaccia commented:
Not sure I understand your question @csgui . Why should a compile-time error be thrown before traversal?

Acaccia commented:
@hugocaillard did a few tests, and it seems some contracts can be deployed if the typechecking doesn't detect the `ArgumentLengthError`. For example, this can deploy:
```clarity
(define-map squares { x: int } { square: int })

(define-public (insert)
  (begin
    (map-insert squares { x: 255 } { square: 2147483648 } { square: 1 })
    (ok true)
  )
)
```

@BowTiedWoo I would recommend you change all the 
```rust
return Err(GeneratorError::ArgumentLengthError(format!(
                "... expected at least 2 arguments, got {}",
                args.len()
            )));
```
by the generation of a wasm runtime error.

@csgui has a lot of experience with runtime errors if you need any help.

BowTiedWoo commented:
> Just adding more context: not sure if the `ArgumentLengthError` will be caught since a compile error is thrown before the traverse step.
> 
> @Acaccia what is your take on that? It seems that this PR should be handled after the work on compile-time error handling, issue #421.

Hi, I've updated the tests and saw that they indeed do not always return the error I've added in traverse. I followed where the error was coming from and went into the `type_checker` in `stacks-core` to understand what was happening.

From here I've made some modifications for the `type-checker` to throw a compile time error for the functions that type_checker was not working as expected and tested them.

I opened a [draft PR](https://github.com/stacks-network/stacks-core/pull/5323) for this.

I wasn’t sure if it would create any backwards incompatibility, but from your message, @Acaccia, I understand that I shouldn’t modify on the `stacks-core` typechecker and only check runtime on our side. I will sync with @csgui when he has some time, I think that would be helpful for me in having the broader picture here.

BowTiedWoo commented:
Makes sense, thank you. Will start looking into it.

Acaccia commented on clar2wasm/src/words/blockinfo.rs:
```diff
@@ -18,6 +36,20 @@ impl ComplexWord for GetBlockInfo {
+                .global_set(get_global(&generator.module, "runtime-error-arg-len")?)
+                .i32_const(ErrorMap::ArgumentCountMismatch as i32)
+                .call(generator.func_by_name("stdlib.runtime-error"))
+                // To avoid having to generate correct return values
+                .unreachable();
```
You will generate correct return values since you don't return after the `unreachable`.

BowTiedWoo replied:
Understood, I deleted it.

Acaccia commented on clar2wasm/src/wasm_generator.rs:
```diff
@@ -110,6 +110,7 @@ pub enum GeneratorError {
     NotImplemented,
     InternalError(String),
     TypeError(String),
+    ArgumentLengthError(String),
```
Could we have something easier to work with than a string here? A string forces us to work with regexes when we want to process the error.
We would need two `u32` for the expected and actual length. Then you can store in memory the 4 bytes of expected followed by the 4 bytes of actual.

BowTiedWoo replied:
This error was no longer used, however the ones that are used are in `error_mapping.rs`.

They are the `ArgumentCountMismatch`, `ArgumentCountAtLeast` and `ArgumentCountAtMost` errors.

Acaccia commented on clar2wasm/src/words/blockinfo.rs:
```diff
@@ -19,10 +37,17 @@ impl ComplexWord for GetBlockInfo {
-                "get-block-info? expected 2 arguments, got {}",
-                args.len()
-            )));
+            let (arg_name_offset, arg_name_len) =
+                generator.add_string_literal(&format!("expected: {} got: {}", 2, args.len()))?;
+            builder
+                .i32_const(arg_name_offset as i32)
+                .global_set(get_global(&generator.module, "runtime-error-arg-offset")?)
+                .i32_const(arg_name_len as i32)
+                .global_set(get_global(&generator.module, "runtime-error-arg-len")?)
+                .i32_const(ErrorMap::ArgumentCountMismatch as i32)
+                .call(generator.func_by_name("stdlib.runtime-error"))
+                // To avoid having to generate correct return values
+                .unreachable();
```
Yep, that's what you should do, nice job!

BowTiedWoo replied:
Thanks!

Acaccia reviewed (commented):
Nice job, now to replicate it to all functions.........

csgui commented:
@BowTiedWoo, it might be better to update your branch with the latest `main`. There are some structural changes to error_mapping.rs that will affect (and simplify) your work.

csgui commented on clar2wasm/src/error_mapping.rs:
```diff
@@ -257,6 +269,70 @@ fn from_runtime_error_code(
+
+        }
+        ErrorMap::ArgumentCountMismatch => {
+            let runtime_error_arg_offset =
+                get_global_i32(&instance, &mut store, "runtime-error-arg-offset");
+            let runtime_error_arg_len =
+                get_global_i32(&instance, &mut store, "runtime-error-arg-len");
+
+            let memory = instance
+                .get_memory(&mut store, "memory")
+                .unwrap_or_else(|| panic!("Could not find wasm instance memory"));
+            let arg_lengths = read_identifier_from_wasm(
+                memory,
+                &mut store,
+                runtime_error_arg_offset,
+                runtime_error_arg_len,
+            )
+            .unwrap_or_else(|e| panic!("Could not recover arg_name: {e}"));
+
+            let (expected, got) = extract_expected_and_got(&arg_lengths);
```
This code is being reused in other parts of the code. Could you please refactor it into a separate function? Check the work done in the `short_return_value` function and try to replicate the approach here.

BowTiedWoo replied:
To get rid of the duplication here I created a function `get_runtime_error_arg_lengths` based on the `short_return_value` function.

csgui commented on clar2wasm/src/error_mapping.rs:
```diff
@@ -257,6 +269,70 @@ fn from_runtime_error_code(
+            Error::Unchecked(CheckErrors::IncorrectArgumentCount(expected, got))
+        }
+        ErrorMap::ArgumentCountAtLeast => {
+            let runtime_error_arg_offset =
+                get_global_i32(&instance, &mut store, "runtime-error-arg-offset");
+            let runtime_error_arg_len =
+                get_global_i32(&instance, &mut store, "runtime-error-arg-len");
+
+            let memory = instance
+                .get_memory(&mut store, "memory")
+                .unwrap_or_else(|| panic!("Could not find wasm instance memory"));
+            let arg_lengths = read_identifier_from_wasm(
+                memory,
+                &mut store,
+                runtime_error_arg_offset,
+                runtime_error_arg_len,
+            )
+            .unwrap_or_else(|e| panic!("Could not recover arg_name: {e}"));
+
+            let (expected, got) = extract_expected_and_got(&arg_lengths);
```
Same comment above about code duplication.

csgui commented on clar2wasm/src/error_mapping.rs:
```diff
@@ -257,6 +269,70 @@ fn from_runtime_error_code(
+            Error::Unchecked(CheckErrors::RequiresAtLeastArguments(expected, got))
+        }
+        ErrorMap::ArgumentCountAtMost => {
+            let runtime_error_arg_offset =
+                get_global_i32(&instance, &mut store, "runtime-error-arg-offset");
+            let runtime_error_arg_len =
+                get_global_i32(&instance, &mut store, "runtime-error-arg-len");
+
+            let memory = instance
+                .get_memory(&mut store, "memory")
+                .unwrap_or_else(|| panic!("Could not find wasm instance memory"));
+            let arg_lengths = read_identifier_from_wasm(
+                memory,
+                &mut store,
+                runtime_error_arg_offset,
+                runtime_error_arg_len,
+            )
+            .unwrap_or_else(|e| panic!("Could not recover arg_name: {e}"));
+
+            let (expected, got) = extract_expected_and_got(&arg_lengths);
```
Same comment above about code duplication.

csgui commented on clar2wasm/src/words/bindings.rs:
```diff
@@ -1,8 +1,26 @@
 use crate::wasm_generator::{ArgumentsExt, GeneratorError, WasmGenerator};
 use crate::words::ComplexWord;
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
This function is being reused in several parts of the code. Could you please move it to `wasm_utils.rs`? I don't usually prefer this "sweeping functions under the rug" approach, but it’s better than duplicating code.

BowTiedWoo replied:
I moved the `get_global` function in `wasm_utils.rs`.

I am now using it by importing the function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/blockinfo.rs:
```diff
@@ -1,8 +1,26 @@
+use crate::error_mapping::ErrorMap;
 use crate::wasm_generator::{ArgumentsExt, GeneratorError, WasmGenerator};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/conditionals.rs:
```diff
@@ -11,6 +11,22 @@ use crate::wasm_generator::{
 };
 use crate::words;
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
+
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/consensus_buff.rs:
```diff
@@ -1,12 +1,30 @@
     WasmGenerator,
 };
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
+
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/constants.rs:
```diff
@@ -1,10 +1,27 @@
 use crate::wasm_generator::{ArgumentsExt, GeneratorError, WasmGenerator};
 use crate::wasm_utils::{get_type_size, is_in_memory_type};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
+
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/contract.rs:
```diff
@@ -3,11 +3,28 @@ use clarity::vm::types::signatures::CallableSubtype;
+use crate::error_mapping::ErrorMap;
 use crate::wasm_generator::{ArgumentsExt, GeneratorError, WasmGenerator};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/control_flow.rs:
```diff
@@ -1,11 +1,28 @@
 use crate::error_mapping::ErrorMap;
 use crate::wasm_generator::{drop_value, ArgumentsExt, GeneratorError, WasmGenerator};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/data_vars.rs:
```diff
@@ -1,10 +1,27 @@
+use crate::error_mapping::ErrorMap;
 use crate::wasm_generator::{ArgumentsExt, GeneratorError, LiteralMemoryEntry, WasmGenerator};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/default_to.rs:
```diff
@@ -1,12 +1,30 @@
     clar2wasm_ty, drop_value, ArgumentsExt, GeneratorError, WasmGenerator,
 };
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/enums.rs:
```diff
@@ -1,11 +1,29 @@
     add_placeholder_for_type, clar2wasm_ty, ArgumentsExt, GeneratorError, WasmGenerator,
 };
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/equal.rs:
```diff
@@ -2,13 +2,30 @@ use clarity::vm::types::signatures::CallableSubtype;
     clar2wasm_ty, drop_value, ArgumentsExt, GeneratorError, SequenceElementType, WasmGenerator,
 };
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/functions.rs:
```diff
@@ -1,8 +1,26 @@
+use crate::error_mapping::ErrorMap;
 use crate::wasm_generator::{ArgumentsExt, FunctionKind, GeneratorError, WasmGenerator};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/options.rs:
```diff
@@ -35,6 +37,22 @@ pub fn traverse_optional(
     Ok(())
 }
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/principal.rs:
```diff
@@ -7,13 +7,30 @@ use clarity::{
     add_placeholder_for_clarity_type, clar2wasm_ty, ArgumentsExt, GeneratorError, WasmGenerator,
 };
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/print.rs:
```diff
@@ -1,10 +1,28 @@
 use crate::wasm_generator::{ArgumentsExt, GeneratorError, WasmGenerator};
 use crate::wasm_utils::signature_from_string;
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/responses.rs:
```diff
@@ -38,6 +40,22 @@ pub fn traverse_response(
     Ok(())
 }
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/secp256k1.rs:
```diff
@@ -1,8 +1,26 @@
+use crate::error_mapping::ErrorMap;
 use crate::wasm_generator::{ArgumentsExt, GeneratorError, WasmGenerator};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/sequences.rs:
```diff
@@ -13,6 +13,22 @@ use crate::wasm_generator::{
 };
 use crate::words::{self, ComplexWord};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/stx.rs:
```diff
@@ -1,9 +1,27 @@
+use crate::error_mapping::ErrorMap;
 use crate::wasm_generator::{ArgumentsExt, GeneratorError, WasmGenerator};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/traits.rs:
```diff
@@ -1,8 +1,26 @@
+use crate::error_mapping::ErrorMap;
 use crate::wasm_generator::{ArgumentsExt, GeneratorError, WasmGenerator};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui commented on clar2wasm/src/words/tuples.rs:
```diff
@@ -2,10 +2,28 @@ use std::collections::BTreeMap;
+use crate::error_mapping::ErrorMap;
 use crate::wasm_generator::{clar2wasm_ty, drop_value, GeneratorError, WasmGenerator};
 
+fn get_global(module: &Module, name: &str) -> Result<GlobalId, GeneratorError> {
+    module
+        .globals
+        .iter()
+        .find(|global| {
+            global
+                .name
+                .as_ref()
+                .map_or(false, |other_name| name == other_name)
+        })
+        .map(|global| global.id())
+        .ok_or_else(|| {
+            GeneratorError::InternalError(format!("Expected to find a global named ${name}"))
+        })
+}
```
Reuse this function from `wasm_utils.rs`.

csgui reviewed (changes requested):
@BowTiedWoo Good work on this PR. However, with the proposed changes, it’s becoming quite large to review. Let’s adjust it step by step. Please check the comments I’ve left.

BowTiedWoo commented:
Thank you @csgui!

I removed all the duplications with your suggestions from the code.

csgui commented:
Ok. Nice! 

Now consider removing the duplication on the arguments check for each `word` implementation.

For instance you can use something like (code not tested)

```rust
fn check_argument_count(
    generator: &mut WasmGenerator,
    builder: &mut walrus::InstrSeqBuilder,
    expected: usize,
    actual: usize,
) -> Result<(), GeneratorError> {
    if actual != expected {
        let (arg_name_offset_start, arg_name_len_expected) =
            generator.add_literal(&clarity::vm::Value::UInt(expected as u128))?;
        let (_, arg_name_len_got) =
            generator.add_literal(&clarity::vm::Value::UInt(actual as u128))?;
        builder
            .i32_const(arg_name_offset_start as i32)
            .global_set(get_global(&generator.module, "runtime-error-arg-offset")?)
            .i32_const((arg_name_len_expected + arg_name_len_got) as i32)
            .global_set(get_global(&generator.module, "runtime-error-arg-len")?)
            .i32_const(ErrorMap::ArgumentCountMismatch as i32)
            .call(generator.func_by_name("stdlib.runtime-error"));
        Err(GeneratorError::ArgumentCountMismatch(expected, actual))
    } else {
        Ok(())
    }
}
```

and reuse it on `traverse` functions, like

```rust
fn traverse(
    &self,
    generator: &mut WasmGenerator,
    builder: &mut walrus::InstrSeqBuilder,
    _expr: &SymbolicExpression,
    args: &[SymbolicExpression],
) -> Result<(), GeneratorError> {
    check_argument_count(generator, builder, [1 | 2 | 3 | etc...], args.len());
    ...
```

csgui commented:
@BowTiedWoo Great work so far! 👍  Now it's necessary to validate all the tests you've written where the type checker is failing and the runtime error is being thrown. You can check the example below and use it to guide your work. Let me know if you have any question. Thanks for the patience on all work on this issue.

```rust
fn get_block_info_more_than_two_args() {
    // TODO: see issue #488
    // The inconsistency in function arguments should have been caught by the typechecker.
    // The runtime error below is being used as a workaround for a typechecker issue
    // where certain errors are not properly handled.
    // This test should be re-worked once the typechecker is fixed
    // and can correctly detect all argument inconsistencies
    crosscheck(
        "(get-block-info? burnchain-header-hash u0 u0)",
        Err(Error::Unchecked(CheckErrors::IncorrectArgumentCount(2, 3))),
    )
}
```

csgui commented on clar2wasm/src/words/bindings.rs:
```diff
@@ -100,6 +109,17 @@ mod tests {
+    #[test]
+    fn let_less_than_two_args() {
+        let result = evaluate("(let ((current-count (count u1))))");
+        println!("{:#?}", result);
```
Please, remove `println!`s from the code.

BowTiedWoo commented:
Thank you @csgui!

I added comments for the tests where typechecker fails.

csgui reviewed (approved):
Code LGTM! Thanks @BowTiedWoo for the great work on this PR!

Acaccia reviewed (changes requested):
Great job! Two remarks:

- there is an optimization that could be done by short-cutting the traversal of an expression.
- we should also handle the number of arguments of user-defined functions. Look for `WasmGenerator::traverse_call_user_defined`

Acaccia reviewed (commented):
Looks great, just two questions...

BowTiedWoo commented on clar2wasm/src/error_mapping.rs:
```diff
@@ -312,3 +360,32 @@ fn short_return_value(
+    let memory = instance
+        .get_memory(&mut *store, "memory")
+        .unwrap_or_else(|| panic!("Could not find wasm instance memory"));
+    let arg_lengths = read_identifier_from_wasm(
```
This is actually a String that contains the lengths of the expected and actual argument count as bytes.

Acaccia replied:
What is the string like?

BowTiedWoo commented on clar2wasm/src/error_mapping.rs:
```diff
@@ -279,6 +303,30 @@ fn get_global_i32(instance: &Instance, store: &mut impl AsContextMut, name: &str
+fn extract_expected_and_got(arg_lengths: &str) -> (usize, usize) {
+    let bytes = arg_lengths.as_bytes();
+
+    // Assuming the first 4 bytes represent the expected value
+    let expected = u32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]]) as usize;
+
+    // Assuming the next 4 bytes represent the got value
+    let got = u32::from_le_bytes([bytes[16], bytes[17], bytes[18], bytes[19]]) as usize;
```
When the values for the expected and actual lengths are written on the stack, they are written with the  `add_literal` function that takes as argument a `clarity::vm::Value::Uint` as u128 (16 bytes) and I read them from memory as u32s (4 bytes) values.

Acaccia replied:
There is also `add_bytes_literal` if you want to write whatever bytes in memory.

BowTiedWoo replied:
Got it, I changed how I store the expected and actual argument lengths on the stack, I am now using `add_bytes_literal` and for getting the `arg_lengths` i am now using `read_bytes_from_wasm` that returns a vector of bytes. As for the indexes, they are now from 0 to 3 and 4 to 7 (8 bytes)

Acaccia reviewed (approved):
LGTM!

csgui reviewed (approved):
LGTM! Thanks.
